src/include/denc.h

```diff
@@ -0,0 +1,1176 @@
+      enum { bounded = false };
+      enum { featured = false };
+      static void bound_encode(const T &o, size_t& p, uint64_t f=0);
+      static void encode(const T &o, buffer::list::contiguous_appender& p,
+		         uint64_t f=0);
+      static void decode(T& o, buffer::ptr::iterator &p);
+    };
+
+  or (for featured objects)
+
+    template<>
+    struct denc_traits<T> {
+      enum { supported = true };
+      enum { bounded = false };
+      enum { featured = true };
+      static void bound_encode(const T &o, size_t& p, uint64_t f);
+      static void encode(const T &o, buffer::list::contiguous_appender& p,
+		         uint64_t f);
+      static void decode(T& o, buffer::ptr::iterator &p);
+    };
+
+  - denc_traits<T> is normally declared via the WRITE_CLASS_DENC(type) macro,
+  which is used in place of the old-style WRITE_CLASS_ENCODER(type) macro.
+  There are _FEATURED and _BOUNDED variants.  The class traits simply call
+  into class methods of the same name (see below).
+
+  - denc_traits<T> can also be written explicitly for some type to indicate
+  how it should be encoded.  This is the "source of truth" for how a type
+  is encoded.
+
+  - denc_traits<T> are declared for the base integer types, string, bufferptr,
+  and bufferlist base types.
+
+  - denc_traits<std::foo<T>>-like traits are declared for standard container
+  types.
+
+
+  class methods look like so
+  ==========================
+
+    void bound_encode(size_t& p) const;
+    void encode(buffer::list::contiguous_appender& p) const;
+    void decode(buffer::ptr::iterator &p);
+
+  or (for featured objects)
+
+    void bound_encode(size_t& p, uint64_t f) const;
+    void encode(buffer::list::contiguous_appender& p, uint64_t f) const;
+    void decode(buffer::ptr::iterator &p);
+
+  - These are normally invoked by the denc_traits<> methods that are
+  declared via WRITE_CLASS_DENC, although you can also invoke them explicitly
+  in your code.
+
+  - These can be defined either explicitly (as above), or can be "magically"
+  defined all in one go using the DENC macro and DENC_{START,FINISH} helpers
+  (which work like the legacy {ENCODE,DECODE}_{START,FINISH} macros):
+
+    class foo_t {
+      ...
+      DENC(foo_t, v, p) {
+        DENC_START(1, 1, p);
+        denc(v.foo, p);
+        denc(v.bar, p);
+        denc(v.baz, p);
+        DENC_FINISH(p);
+      }
+      ...
+    };
+    WRITE_CLASS_DENC(foo_t)
+
+  */
+
+
+// -----------------------------------------------------------------------
+// base types
+
+// Just like the existing encode/decode machinery. The environment
+// provides a rich set of pre-defined encodes for primitive types and
+// containers
+
+// These macros all handle byte-ordering automatically for you :) All
+// external serializations are in little endian order
+
+// itype == internal type
+// otype == external type, i.e., the type on the wire
+
+// NOTE: set supported == 2 instead of true.  This prevents these from
```

Comment:
It would be useful to have a place in the comments where the values of the supported, featured, etc. are explicitly documented in a table.